Build OpenBLAS 0.3.6 for iOS

Hello! I'm trying to build Open BLAS `0.3.5` version for iOS. Here's my shell script I'm using for building:

```
TOOLCHAIN_PATH=/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
SYSROOT_PATH=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk
make TARGET=ARMV8 BINARY=64 HOSTCC=clang CC="$TOOLCHAIN_PATH/clang -isysroot $SYSROOT_PATH -arch arm64 -miphoneos-version-min=10.0 -O2" NOFORTRAN=1 libs
```

When I execute it, I'm happen to have this error:

```
clang: error: unknown argument: '-ru'
clang: error: no such file or directory: '../libopenblas_armv8p-r0.3.6.a'
clang: warning: no such sysroot directory: '/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/-ar' [-Wmissing-sysroot]
```

How can this be fixed? Thanks!

#569 I found a fix for that here. But now when I try to build my XCode project with new library, I have those errors:

ld: warning: directory not found for option '-L/Users/user/Documents/Work/AlgofaceWork/algoface-ios-209-landmarks-tracker/Algoface-Landmarks-Tracker/tracker/3rdparty/eigen/OpenBLAS_iOS_0.2.21'
```
Undefined symbols for architecture arm64:
  "_sdot_k", referenced from:
      _strmv_TLN in libopenblas.a(strmv_TLN.o)
      _strmv_TLU in libopenblas.a(strmv_TLU.o)
      _strmv_TUN in libopenblas.a(strmv_TUN.o)
      _strmv_TUU in libopenblas.a(strmv_TUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TUN.o)
      ...
  "_scopy_k", referenced from:
      _strmv_NLN in libopenblas.a(strmv_NLN.o)
      _strmv_NLU in libopenblas.a(strmv_NLU.o)
      _strmv_NUN in libopenblas.a(strmv_NUN.o)
      _strmv_NUU in libopenblas.a(strmv_NUU.o)
      _strmv_TLN in libopenblas.a(strmv_TLN.o)
      _strmv_TLU in libopenblas.a(strmv_TLU.o)
      _strmv_TUN in libopenblas.a(strmv_TUN.o)
      ...
  "_strmm_kernel_RN", referenced from:
      _strmm_RNUN in libopenblas.a(strmm_RNUN.o)
      _strmm_RNUU in libopenblas.a(strmm_RNUU.o)
      _strmm_RTLN in libopenblas.a(strmm_RTLN.o)
      _strmm_RTLU in libopenblas.a(strmm_RTLU.o)
  "_strmm_kernel_LN", referenced from:
      _strmm_LNUN in libopenblas.a(strmm_LNUN.o)
      _strmm_LNUU in libopenblas.a(strmm_LNUU.o)
      _strmm_LTLN in libopenblas.a(strmm_LTLN.o)
      _strmm_LTLU in libopenblas.a(strmm_LTLU.o)
  "_strmm_kernel_LT", referenced from:
      _strmm_LNLN in libopenblas.a(strmm_LNLN.o)
      _strmm_LNLU in libopenblas.a(strmm_LNLU.o)
      _strmm_LTUN in libopenblas.a(strmm_LTUN.o)
      _strmm_LTUU in libopenblas.a(strmm_LTUU.o)
  "_sscal_k", referenced from:
      _sgemv_ in libopenblas.a(sgemv.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NUN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLU.o)
      ...
  "_sgemm_kernel", referenced from:
      _sgemm_nn in libopenblas.a(sgemm_nn.o)
      _sgemm_nt in libopenblas.a(sgemm_nt.o)
      _inner_thread in libopenblas.a(sgemm_thread_nn.o)
      _inner_thread in libopenblas.a(sgemm_thread_nt.o)
      _inner_thread in libopenblas.a(sgemm_thread_tn.o)
      _inner_thread in libopenblas.a(sgemm_thread_tt.o)
      _sgemm_tn in libopenblas.a(sgemm_tn.o)
      ...
  "_strmm_kernel_RT", referenced from:
      _strmm_RNLN in libopenblas.a(strmm_RNLN.o)
      _strmm_RNLU in libopenblas.a(strmm_RNLU.o)
      _strmm_RTUN in libopenblas.a(strmm_RTUN.o)
      _strmm_RTUU in libopenblas.a(strmm_RTUU.o)
  "_sgemv_n", referenced from:
      l___const.sgemv_.gemv in libopenblas.a(sgemv.o)
      _gemv_kernel in libopenblas.a(sgemv_thread_n.o)
      _strmv_NLN in libopenblas.a(strmv_NLN.o)
      _strmv_NLU in libopenblas.a(strmv_NLU.o)
      _strmv_NUN in libopenblas.a(strmv_NUN.o)
      _strmv_NUU in libopenblas.a(strmv_NUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLN.o)
      ...
  "_sgemv_t", referenced from:
      l___const.sgemv_.gemv in libopenblas.a(sgemv.o)
      _gemv_kernel in libopenblas.a(sgemv_thread_t.o)
      _strmv_TLN in libopenblas.a(strmv_TLN.o)
      _strmv_TLU in libopenblas.a(strmv_TLU.o)
      _strmv_TUN in libopenblas.a(strmv_TUN.o)
      _strmv_TUU in libopenblas.a(strmv_TUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLN.o)
      ...
     (maybe you meant: _sgemv_thread_n, _sgemv_thread_t )
  "_saxpy_k", referenced from:
      _saxpy_ in libopenblas.a(saxpy.o)
      _strmv_NLN in libopenblas.a(strmv_NLN.o)
      _strmv_NLU in libopenblas.a(strmv_NLU.o)
      _strmv_NUN in libopenblas.a(strmv_NUN.o)
      _strmv_NUU in libopenblas.a(strmv_NUU.o)
      _strmv_thread_NLN in libopenblas.a(strmv_thread_NLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLN.o)
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Can that be an issue that XCode can't compile assembly files? 

I think it's pretty significant moment, because I forced XCode to build C versions of functions previously and the library is not fast as I think it should be

I removed `\@` stuff and now I have such errors:

```
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
/usr/local/opt/llvm/bin/clang -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -arch arm64 -miphoneos-version-min=8.0 -O2 -c -O2 -DMAX_STACK_ALLOC=2048 -Wall -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=8 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.5\" -march=armv8-a -DASMNAME=_ssymv_L -DASMFNAME=_ssymv_L_ -DNAME=ssymv_L_ -DCNAME=ssymv_L -DCHAR_NAME=\"ssymv_L_\" -DCHAR_CNAME=\"ssymv_L\" -DNO_AFFINITY -I.. -UDOUBLE  -UCOMPLEX -UCOMPLEX -UDOUBLE -DLOWER ../kernel/arm64/../generic/symv_k.c -o ssymv_L.o
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: /usr/local/opt/llvm/bin/clang -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -arch arm64 -miphoneos-version-min=8.0 -O2 -c -O2 -DMAX_STACK_ALLOC=2048 -Wall -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=8 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.5\" -march=armv8-a -DASMNAME=_sger_k -DASMFNAME=_sger_k_ -DNAME=sger_k_ -DCNAME=sger_k -DCHAR_NAME=\"sger_k_\" -DCHAR_CNAME=\"sger_k\" -DNO_AFFINITY -I.. -UDOUBLE  -UCOMPLEX -UDOUBLE ../kernel/arm64/../generic/ger.c -o sger_k.o
error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
<instantiation>:14:1: error: invalid symbol redefinition
KERNEL_F1_SCALE_GE_X_:
^
../kernel/arm64/nrm2.S:87:1: error: invalid symbol redefinition
KERNEL_F1_NEXT_:
^
/usr/local/opt/llvm/bin/clang -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -arch arm64 -miphoneos-version-min=8.0 -O2 -O2 -DMAX_STACK_ALLOC=2048 -Wall -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=8 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.5\" -march=armv8-a -DASMNAME=_sgemm_kernel -DASMFNAME=_sgemm_kernel_ -DNAME=sgemm_kernel_ -DCNAME=sgemm_kernel -DCHAR_NAME=\"sgemm_kernel_\" -DCHAR_CNAME=\"sgemm_kernel\" -DNO_AFFINITY -I.. -UDOUBLE  -UCOMPLEX -c -UDOUBLE -UCOMPLEX ../kernel/arm64/sgemm_kernel_16x4.S -o sgemm_kernel.o
make[1]: *** [snrm2_k.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [libs] Error 1
MacBook-Pro-Mojo:OpenBLAS-release_0.3.5 user$ 

```

Do you have any comments on how it can be fixed in Assembly?

Hello @brada4!

Now my `openBlasBuild.sh` looks like this:

```
TOOLCHAIN_PATH=/usr/bin
SYSROOT_PATH=/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk
make TARGET=ARMV8 BINARY=64 HOSTCC=gcc CC="$TOOLCHAIN_PATH/clang -isysroot $SYSROOT_PATH -arch arm64 -miphoneos-version-min=8.0 -O2 -fno-integrated-as" NOFORTRAN=1 libs
```

I ran `make clean` and launched this script. I got those errors in the output:

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/as: can't specifiy -Q with -arch arm64
clangclang: error: clangassembler command failed with exit code 1 (use -v to see invocation): 
error: assembler command failed with exit code 1 (use -v to see invocation)
clang: error: assembler command failed with exit code 1 (use -v to see invocation)clang
: error: assembler command failed with exit code 1 (use -v to see invocation)
clang: error: assembler command failed with exit code 1 (use -v to see invocation): 
error: assembler command failed with exit code 1 (use -v to see invocation)
clang: error: assembler command failed with exit code 1 (use -v to see invocation)
clang: error: assembler command failed with exit code 1 (use -v to see invocation)
make[1]: *** [sasum.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [dsdot.o] Error 1
make[1]: *** [sdsdot.o] Error 1
make[1]: *** [sdot.o] Error 1
make[1]: *** [sscal.o] Error 1
make[1]: *** [scopy.o] Error 1
make[1]: *** [sswap.o] Error 1
make[1]: *** [saxpy.o] Error 1
make: *** [libs] Error 1
```

@martin-frbg @brada4 Ok, I think I make some progress.

Here's how `nrm2.S` looks like. That's one of files who cause trouble. That's only the beginning of the file because I think that part matters now:

```
.macro KERNEL_F1
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_\@
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_\@
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_\@
KERNEL_F1_SCALE_GE_X_\@:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_\@
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_\@
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_\@
KERNEL_F1_SCALE_GE_X_\@:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_\@:
.endm

.macro KERNEL_S1
#if !defined(DOUBLE)
	ldr	s4, [X]
	fcmp	s4, REGZERO
	beq	KERNEL_S1_NEXT
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_S1_SCALE_GE_X
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_S1_NEXT
KERNEL_S1_SCALE_GE_X:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X]
	fcmp	d4, REGZERO
	beq	KERNEL_S1_NEXT
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_S1_SCALE_GE_X
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_S1_NEXT
KERNEL_S1_SCALE_GE_X:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_S1_NEXT:
	add	X, X, INC_X
.endm

.macro KERNEL_F8
	KERNEL_F1
	KERNEL_F1
	KERNEL_F1
	KERNEL_F1
	KERNEL_F1
	KERNEL_F1
	KERNEL_F1
	KERNEL_F1
.endm
```

Here's how I changed it. @martin-frbg please verify if I understood your advice correctly:

```
.macro KERNEL_F1_0
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_0
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_0
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_0
KERNEL_F1_SCALE_GE_X_0:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_0
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_0
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_0
KERNEL_F1_SCALE_GE_X_0:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_0:
.endm

.macro KERNEL_F1_1
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_1
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_1
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_1
KERNEL_F1_SCALE_GE_X_1:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_1
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_1
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_1
KERNEL_F1_SCALE_GE_X_1:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_1:
.endm

.macro KERNEL_F1_2
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_2
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_2
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_2
KERNEL_F1_SCALE_GE_X_2:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_2
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_2
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_2
KERNEL_F1_SCALE_GE_X_2:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_2:
.endm

.macro KERNEL_F1_3
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_3
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_3
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_3
KERNEL_F1_SCALE_GE_X_3:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_3
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_3
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_3
KERNEL_F1_SCALE_GE_X_3:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_3:
.endm

.macro KERNEL_F1_4
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_4
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_4
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_4
KERNEL_F1_SCALE_GE_X_4:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_4
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_4
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_4
KERNEL_F1_SCALE_GE_X_4:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_4:
.endm

.macro KERNEL_F1_5
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_5
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_5
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_5
KERNEL_F1_SCALE_GE_X_5:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_5
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_5
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_5
KERNEL_F1_SCALE_GE_X_5:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_5:
.endm

.macro KERNEL_F1_6
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_6
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_6
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_6
KERNEL_F1_SCALE_GE_X_6:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_6
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_6
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_6
KERNEL_F1_SCALE_GE_X_6:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_6:
.endm

.macro KERNEL_F1_7
#if !defined(DOUBLE)
	ldr	s4, [X], #4
	fcmp	s4, REGZERO
	beq	KERNEL_F1_NEXT_7
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_F1_SCALE_GE_X_7
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_F1_NEXT_7
KERNEL_F1_SCALE_GE_X_7:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X], #8
	fcmp	d4, REGZERO
	beq	KERNEL_F1_NEXT_7
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_F1_SCALE_GE_X_7
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_F1_NEXT_7
KERNEL_F1_SCALE_GE_X_7:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_F1_NEXT_7:
.endm

.macro KERNEL_S1
#if !defined(DOUBLE)
	ldr	s4, [X]
	fcmp	s4, REGZERO
	beq	KERNEL_S1_NEXT
	fabs	s4, s4
	fcmp	SCALE, s4
	bge	KERNEL_S1_SCALE_GE_X
	fdiv	s2, SCALE, s4
	fmul	s2, s2, s2
	fmul	s3, SSQ, s2
	fadd	SSQ, REGONE, s3
	fmov	SCALE, s4
	b	KERNEL_S1_NEXT
KERNEL_S1_SCALE_GE_X:
	fdiv	s2, s4, SCALE
	fmla	SSQ, s2, v2.s[0]
#else
	ldr	d4, [X]
	fcmp	d4, REGZERO
	beq	KERNEL_S1_NEXT
	fabs	d4, d4
	fcmp	SCALE, d4
	bge	KERNEL_S1_SCALE_GE_X
	fdiv	d2, SCALE, d4
	fmul	d2, d2, d2
	fmul	d3, SSQ, d2
	fadd	SSQ, REGONE, d3
	fmov	SCALE, d4
	b	KERNEL_S1_NEXT
KERNEL_S1_SCALE_GE_X:
	fdiv	d2, d4, SCALE
	fmla	SSQ, d2, v2.d[0]
#endif
KERNEL_S1_NEXT:
	add	X, X, INC_X
.endm

.macro KERNEL_F8
	KERNEL_F1_0
	KERNEL_F1_1
	KERNEL_F1_2
	KERNEL_F1_3
	KERNEL_F1_4
	KERNEL_F1_5
	KERNEL_F1_6
	KERNEL_F1_7
.endm
```

When I try to compile the library with those changes, I get next error. That's the whole error log:

```
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/nrm2-c640f1.s:306:2: error: unrecognized instruction mnemonic
 KERNEL_F1
 ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -arch arm64 -miphoneos-version-min=8.0 -O2 -c -O2 -DMAX_STACK_ALLOC=2048 -Wall -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=8 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.6\" -march=armv8-a -DASMNAME=_srot_k -DASMFNAME=_srot_k_ -DNAME=srot_k_ -DCNAME=srot_k -DCHAR_NAME=\"srot_k_\" -DCHAR_CNAME=\"srot_k\" -DNO_AFFINITY -I.. -UDOUBLE  -UCOMPLEX -UCOMPLEX -UCOMPLEX -UDOUBLE  ../kernel/arm64/rot.S -o srot_k.o
make[1]: *** [snrm2_k.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [libs] Error 1

```

I see that `KERNEL_F1` is also used here in `nrm2.S`:

```
.Lnrm2_kernel_F10:

	KERNEL_F1

	subs	I, I, #1
	bne	.Lnrm2_kernel_F10

	b	.Lnrm2_kernel_L999
```

Error is pretty straight-forward. Compiler can't find `KERNEL_F1` because there's none, there're new 8 `KERNEL_F1_[number]` macroses. 

What should I do to fix that? Should I create another copy of `KERNEL_F1` in order to use here?

I understand that I'm solving the problem with an "axe". But this approach I can at least understand at that point :)

I did the same trick for `znrm2.S`. Now my errors look like this:

```
<instantiation>:4:21: error: unexpected token in argument list
 beq KERNEL_S1_NEXT_\@
                    ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:7:28: error: unexpected token in argument list
 bge KERNEL_S1_SCALE_GE_XR_\@
                           ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:13:19: error: unexpected token in argument list
 b KERNEL_S1_NEXT_\@
                  ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:14:23: error: unknown token in expression
KERNEL_S1_SCALE_GE_XR_\@:
                      ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:14:23: error: invalid operand
KERNEL_S1_SCALE_GE_XR_\@:
                      ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:17:16: error: unknown token in expression
KERNEL_S1_NEXT_\@:
               ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:17:16: error: invalid operand
KERNEL_S1_NEXT_\@:
               ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:20:20: error: unexpected token in argument list
 beq KERNEL_S1_END_\@
                   ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:23:28: error: unexpected token in argument list
 bge KERNEL_S1_SCALE_GE_XI_\@
                           ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:29:18: error: unexpected token in argument list
 b KERNEL_S1_END_\@
                 ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:30:23: error: unknown token in expression
KERNEL_S1_SCALE_GE_XI_\@:
                      ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
<instantiation>:30:23: error: invalid operand
KERNEL_S1_SCALE_GE_XI_\@:
                      ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
../kernel/arm64/znrm2.S:740:15: error: unknown token in expression
KERNEL_S1_END_\@:
              ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
../kernel/arm64/znrm2.S:740:15: error: invalid operand
KERNEL_S1_END_\@:
              ^
/var/folders/xt/7v6bxk2n19zft1y3yzymdc580000gq/T/znrm2-f2371d.s:503:2: note: while in macro instantiation
 KERNEL_S1
 ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS.sdk -arch arm64 -miphoneos-version-min=8.0 -O2 -c -O2 -DMAX_STACK_ALLOC=2048 -Wall -DF_INTERFACE_GFORT -fPIC -DNO_LAPACK -DNO_LAPACKE -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=8 -DMAX_PARALLEL_NUMBER=1 -DVERSION=\"0.3.6\" -march=armv8-a -DASMNAME=_caxpby_k -DASMFNAME=_caxpby_k_ -DNAME=caxpby_k_ -DCNAME=caxpby_k -DCHAR_NAME=\"caxpby_k_\" -DCHAR_CNAME=\"caxpby_k\" -DNO_AFFINITY -I.. -UDOUBLE  -DCOMPLEX -DCOMPLEX -UCONJ -UDOUBLE ../kernel/arm64/../arm/zaxpby.c -o caxpby_k.o
make[1]: *** [cnrm2_k.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make: *** [libs] Error 1
```

What's suspicious to me here is that this macro is used only here in `znrm2.S`:

```
.Lznrm2_kernel_S10:

	KERNEL_S1

	subs	I, I, #1
	bne	.Lznrm2_kernel_S10
```

Now it makes me thinking - how often this `.Lznrm2_kernel_S10` gets called? Maybe this automatic labels counter important after all. 

Or was it done this way just to keep coding style the same?

@martin-frbg Here're modified `nrm2.S` and `znm2.S` for `0.3.6` version. With those library successfuly compiles 

[Archive.zip](https://github.com/xianyi/OpenBLAS/files/3675932/Archive.zip)


But, when I try to compile the project with new the library I have following errors:

```
Undefined symbols for architecture arm64:
  "_sdot_k", referenced from:
      _strmv_TLN in libopenblas.a(strmv_TLN.o)
      _strmv_TLU in libopenblas.a(strmv_TLU.o)
      _strmv_TUN in libopenblas.a(strmv_TUN.o)
      _strmv_TUU in libopenblas.a(strmv_TUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TUN.o)
      ...
  "_scopy_k", referenced from:
      _strmv_NLN in libopenblas.a(strmv_NLN.o)
      _strmv_NLU in libopenblas.a(strmv_NLU.o)
      _strmv_NUN in libopenblas.a(strmv_NUN.o)
      _strmv_NUU in libopenblas.a(strmv_NUU.o)
      _strmv_TLN in libopenblas.a(strmv_TLN.o)
      _strmv_TLU in libopenblas.a(strmv_TLU.o)
      _strmv_TUN in libopenblas.a(strmv_TUN.o)
      ...
  "_strmm_kernel_RN", referenced from:
      _strmm_RNUN in libopenblas.a(strmm_RNUN.o)
      _strmm_RNUU in libopenblas.a(strmm_RNUU.o)
      _strmm_RTLN in libopenblas.a(strmm_RTLN.o)
      _strmm_RTLU in libopenblas.a(strmm_RTLU.o)
  "_strmm_kernel_LN", referenced from:
      _strmm_LNUN in libopenblas.a(strmm_LNUN.o)
      _strmm_LNUU in libopenblas.a(strmm_LNUU.o)
      _strmm_LTLN in libopenblas.a(strmm_LTLN.o)
      _strmm_LTLU in libopenblas.a(strmm_LTLU.o)
  "_strmm_kernel_LT", referenced from:
      _strmm_LNLN in libopenblas.a(strmm_LNLN.o)
      _strmm_LNLU in libopenblas.a(strmm_LNLU.o)
      _strmm_LTUN in libopenblas.a(strmm_LTUN.o)
      _strmm_LTUU in libopenblas.a(strmm_LTUU.o)
  "_sscal_k", referenced from:
      _sgemv_ in libopenblas.a(sgemv.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NUN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLU.o)
      ...
  "_sgemm_kernel", referenced from:
      _sgemm_nn in libopenblas.a(sgemm_nn.o)
      _sgemm_nt in libopenblas.a(sgemm_nt.o)
      _inner_thread in libopenblas.a(sgemm_thread_nn.o)
      _inner_thread in libopenblas.a(sgemm_thread_nt.o)
      _inner_thread in libopenblas.a(sgemm_thread_tn.o)
      _inner_thread in libopenblas.a(sgemm_thread_tt.o)
      _sgemm_tn in libopenblas.a(sgemm_tn.o)
      ...
  "_strmm_kernel_RT", referenced from:
      _strmm_RNLN in libopenblas.a(strmm_RNLN.o)
      _strmm_RNLU in libopenblas.a(strmm_RNLU.o)
      _strmm_RTUN in libopenblas.a(strmm_RTUN.o)
      _strmm_RTUU in libopenblas.a(strmm_RTUU.o)
  "_sgemv_n", referenced from:
      l___const.sgemv_.gemv in libopenblas.a(sgemv.o)
      _gemv_kernel in libopenblas.a(sgemv_thread_n.o)
      _strmv_NLN in libopenblas.a(strmv_NLN.o)
      _strmv_NLU in libopenblas.a(strmv_NLU.o)
      _strmv_NUN in libopenblas.a(strmv_NUN.o)
      _strmv_NUU in libopenblas.a(strmv_NUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLN.o)
      ...
  "_sgemv_t", referenced from:
      l___const.sgemv_.gemv in libopenblas.a(sgemv.o)
      _gemv_kernel in libopenblas.a(sgemv_thread_t.o)
      _strmv_TLN in libopenblas.a(strmv_TLN.o)
      _strmv_TLU in libopenblas.a(strmv_TLU.o)
      _strmv_TUN in libopenblas.a(strmv_TUN.o)
      _strmv_TUU in libopenblas.a(strmv_TUU.o)
      _trmv_kernel in libopenblas.a(strmv_thread_TLN.o)
      ...
     (maybe you meant: _sgemv_thread_n, _sgemv_thread_t )
  "_saxpy_k", referenced from:
      _saxpy_ in libopenblas.a(saxpy.o)
      _strmv_NLN in libopenblas.a(strmv_NLN.o)
      _strmv_NLU in libopenblas.a(strmv_NLU.o)
      _strmv_NUN in libopenblas.a(strmv_NUN.o)
      _strmv_NUU in libopenblas.a(strmv_NUU.o)
      _strmv_thread_NLN in libopenblas.a(strmv_thread_NLN.o)
      _trmv_kernel in libopenblas.a(strmv_thread_NLN.o)
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

@ashwinyes here's the log from building the library. I sincerely hope it will tell you guys something:

[mylog.txt](https://github.com/xianyi/OpenBLAS/files/3676809/mylog.txt)

There was only one warning:

```
blas_server.c:585:109: warning: format specifies type 'long' but the argument
      has type 'int' [-Wformat]
  ...failed for thread %ld of %ld: %s\n", i+1,blas_num_threads,msg);
                              ~~~             ^~~~~~~~~~~~~~~~
                              %d
1 warning generated.
```

I don't think that it's a breaker, but anyway worth noticing.

Tried next thing on `dot.S` file:

```
/*******************************************************************************
* End of macro definitions
*******************************************************************************/

    .text
    .align 2
    .globl REALNAME
REALNAME:

	fmov	DOTF, REG0
#if defined(DOUBLE)
```

`nm` output for `sdot_k.o`:

```
libopenblas_armv8p-r0.3.6.a(sdot_k.o):
0000000000000050 t .Ldot_kernel_F1
0000000000000058 t .Ldot_kernel_F10
0000000000000028 t .Ldot_kernel_F4
000000000000001c t .Ldot_kernel_F_BEGIN
00000000000000d8 t .Ldot_kernel_L999
00000000000000bc t .Ldot_kernel_S1
00000000000000c4 t .Ldot_kernel_S10
0000000000000084 t .Ldot_kernel_S4
0000000000000070 t .Ldot_kernel_S_BEGIN
0000000000000000 T _sdot_k
0000000000000000 t ltmp0
```

Function is there. It just must be `;`, there's no other way around.

But simply removing `;` won't do the trick. Apparently new line is important and it's hard to achieve inside a macro.

If I just remove all `;` I will have such error output:

```
../kernel/arm64/amax.S:132:8: error: unexpected token in section switching directive
 .text .align 2 .globl _samax_k _samax_k:
       ^
../kernel/arm64/iamax.S:157:8: error: unexpected token in section switching directive
 .text .align 2 .globl _isamax_k _isamax_k:
       ^
make[1]: *** [samax_k.o] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [isamax_k.o] Error 1
make: *** [libs] Error 1
```

--- a/common_arm64.h
+++ b/common_arm64.h
@@ -103,12 +103,14 @@ static inline int blas_quickdivide(blasint x, blasint y){
 
 #if defined(ASSEMBLER) && !defined(NEEDPARAM)
 
-#define PROLOGUE \
-	.text ;\
-	.align	4 ;\
-	.global	REALNAME ;\
-	.type	REALNAME, %function ;\
+.macro PROLOGUE 
+	.text ;
+	.p2align 2 ;
+	.global	REALNAME ;
+	.type	REALNAME, %function ;
 REALNAME:
+.endm
+
 
 #define EPILOGUE
 

--- a/kernel/arm64/nrm2.S
+++ b/kernel/arm64/nrm2.S
@@ -54,37 +54,38 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #if !defined(DOUBLE)
 	ldr	s4, [X], #4
 	fcmp	s4, REGZERO
-	beq	KERNEL_F1_NEXT_\@
+	beq	2f     /* KERNEL_F1_NEXT_\@ */
+	beq     2f
 	fabs	s4, s4
 	fcmp	SCALE, s4
-	bge	KERNEL_F1_SCALE_GE_X_\@
+	bge	1f     /* KERNEL_F1_SCALE_GE_X_\@ */
 	fdiv	s2, SCALE, s4
 	fmul	s2, s2, s2
 	fmul	s3, SSQ, s2
 	fadd	SSQ, REGONE, s3
 	fmov	SCALE, s4
-	b	KERNEL_F1_NEXT_\@
-KERNEL_F1_SCALE_GE_X_\@:
+	b	2f     /* KERNEL_F1_NEXT_\@ */
+1:               /* KERNEL_F1_SCALE_GE_X_\@: */
 	fdiv	s2, s4, SCALE
 	fmla	SSQ, s2, v2.s[0]
 #else
 	ldr	d4, [X], #8
 	fcmp	d4, REGZERO
-	beq	KERNEL_F1_NEXT_\@
+	beq	2f     /* KERNEL_F1_NEXT_\@ */
 	fabs	d4, d4
 	fcmp	SCALE, d4
-	bge	KERNEL_F1_SCALE_GE_X_\@
+	bge	1f     /* KERNEL_F1_SCALE_GE_X_\@ */
 	fdiv	d2, SCALE, d4
 	fmul	d2, d2, d2
 	fmul	d3, SSQ, d2
 	fadd	SSQ, REGONE, d3
 	fmov	SCALE, d4
-	b	KERNEL_F1_NEXT_\@
-KERNEL_F1_SCALE_GE_X_\@:
+	b	2f     /* KERNEL_F1_NEXT_\@ */
+1:                     /* KERNEL_F1_SCALE_GE_X_\@: */
 	fdiv	d2, d4, SCALE
 	fmla	SSQ, d2, v2.d[0]
 #endif
-KERNEL_F1_NEXT_\@:
+2:                     /* KERNEL_F1_NEXT_\@: */
 .endm
 
 .macro KERNEL_S1

--- a/kernel/arm64/znrm2.S
+++ b/kernel/arm64/znrm2.S
@@ -54,69 +54,69 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #if !defined(DOUBLE)
 	ldr	s4, [X], #4
 	fcmp	s4, REGZERO
-	beq	KERNEL_F1_NEXT_\@
+	beq	2f    /* KERNEL_F1_NEXT_\@ */
 	fabs	s4, s4
 	fcmp	SCALE, s4
-	bge	KERNEL_F1_SCALE_GE_XR_\@
+	bge	1f    /* KERNEL_F1_SCALE_GE_XR_\@ */
 	fdiv	s2, SCALE, s4
 	fmul	s2, s2, s2
 	fmul	s3, SSQ, s2
 	fadd	SSQ, REGONE, s3
 	fmov	SCALE, s4
-	b	KERNEL_F1_NEXT_\@
-KERNEL_F1_SCALE_GE_XR_\@:
+	b	2f    /* KERNEL_F1_NEXT_\@ */
+1:                    /* KERNEL_F1_SCALE_GE_XR_\@: */
 	fdiv	s2, s4, SCALE
 	fmla	SSQ, s2, v2.s[0]
-KERNEL_F1_NEXT_\@:
+2:                    /* KERNEL_F1_NEXT_\@: */
 	ldr	s5, [X], #4
 	fcmp	s5, REGZERO
-	beq	KERNEL_F1_END_\@
+	beq	4f    /* KERNEL_F1_END_\@ */
 	fabs	s5, s5
 	fcmp	SCALE, s5
-	bge	KERNEL_F1_SCALE_GE_XI_\@
+	bge	3f    /* KERNEL_F1_SCALE_GE_XI_\@ */
 	fdiv	s2, SCALE, s5
 	fmul	s2, s2, s2
 	fmul	s3, SSQ, s2
 	fadd	SSQ, REGONE, s3
 	fmov	SCALE, s5
-	b	KERNEL_F1_END_\@
-KERNEL_F1_SCALE_GE_XI_\@:
+	b	4f    /* KERNEL_F1_END_\@ */
+3:                    /* KERNEL_F1_SCALE_GE_XI_\@: */
 	fdiv	s2, s5, SCALE
 	fmla	SSQ, s2, v2.s[0]
 #else
 	ldr	d4, [X], #8
 	fcmp	d4, REGZERO
-	beq	KERNEL_F1_NEXT_\@
+	beq	2f    /* KERNEL_F1_NEXT_\@ */
 	fabs	d4, d4
 	fcmp	SCALE, d4
-	bge	KERNEL_F1_SCALE_GE_XR_\@
+	bge	1f    /* KERNEL_F1_SCALE_GE_XR_\@ */
 	fdiv	d2, SCALE, d4
 	fmul	d2, d2, d2
 	fmul	d3, SSQ, d2
 	fadd	SSQ, REGONE, d3
 	fmov	SCALE, d4
-	b	KERNEL_F1_NEXT_\@
-KERNEL_F1_SCALE_GE_XR_\@:
+	b	2f    /* KERNEL_F1_NEXT_\@ */
+1:                    /* KERNEL_F1_SCALE_GE_XR_\@: */
 	fdiv	d2, d4, SCALE
 	fmla	SSQ, d2, v2.d[0]
-KERNEL_F1_NEXT_\@:
+2:                   /* KERNEL_F1_NEXT_\@: */
 	ldr	d5, [X], #8
 	fcmp	d5, REGZERO
-	beq	KERNEL_F1_END_\@
+	beq	4f  /* KERNEL_F1_END_\@ */
 	fabs	d5, d5
 	fcmp	SCALE, d5
-	bge	KERNEL_F1_SCALE_GE_XI_\@
+	bge	3f    /* KERNEL_F1_SCALE_GE_XI_\@ */
 	fdiv	d2, SCALE, d5
 	fmul	d2, d2, d2
 	fmul	d3, SSQ, d2
 	fadd	SSQ, REGONE, d3
 	fmov	SCALE, d5
-	b	KERNEL_F1_END_\@
-KERNEL_F1_SCALE_GE_XI_\@:
+	b	4f    /* KERNEL_F1_END_\@ */
+3:                    /* KERNEL_F1_SCALE_GE_XI_\@: */
 	fdiv	d2, d5, SCALE
 	fmla	SSQ, d2, v2.d[0]
 #endif
-KERNEL_F1_END_\@:
+4:                    /* KERNEL_F1_END_\@: */
 .endm
 
 .macro KERNEL_S1

